File: dotnet/Core/GameX.Resource/Crytek/FarCry2.cs
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO.Compression;
using System.Linq;

namespace GameX.Crytek;

public static class FarCry2 {
    static readonly Dictionary<string, ZipArchiveEntry> Files;
    static FarCry2() {
        var assembly = typeof(FarCry2).Assembly;
        var s = assembly.GetManifestResourceStream("GameX.Resource.Crytek.FarCry2.zip");
        var arc = new ZipArchive(s, ZipArchiveMode.Read);
        Files = arc.Entries.ToDictionary(s => s.FullName);
    }

    static readonly ConcurrentDictionary<string, Dictionary<ulong, string>> FileHashes = new();
    static readonly ConcurrentDictionary<string, Definition> ObjDefs = new();
    public static Dictionary<ulong, string> GetFileHashes(string path) => FileHashes.GetOrAdd(path, s => Files.TryGetValue(s, out var z) ? FarCryX.HashFilelist32(z) : []);
    public static Definition GetObjDef(string path) => ObjDefs.GetOrAdd(path, s => Files.TryGetValue(s, out var z) ? FarCryX.HashObj32(z) : null);
}

File: python/gamex/Resources/Crytek/FarCry2.py
from zipfile import ZipFile
from importlib import resources
from .FarCryX import hashFilelist32, hashObj32, Definition

f = resources.files().joinpath('FarCry2.zip').open('rb')
arc: ZipFile = ZipFile(f, 'r')
files: dict[str, object] = { s.filename:s for s in arc.infolist() }
fileHashes: dict[str, dict[int, str]] = {}
objDefs: dict[str, Definition] = {}
@staticmethod
def getFileHashes(path: str) -> dict[int, str]:
    if path in fileHashes: return fileHashes[path]
    fileHashes[path] = hashFilelist32(arc, files[path]) if path in files else []
    return fileHashes[path]
@staticmethod
def getObjDef(path: str) -> Definition:
    if path in objDefs: return objDefs[path]
    objDefs[path] = hashObj32(arc, files[path]) if path in files else []
    return objDefs[path]
